### PR TITLE
DDT-752: Conditional check for expectedDepartureDate in POST/PUT

### PR DESCRIPTION
**Pull request by mp-assco**
Dependent on: https://dcsa.atlassian.net/browse/DDT-758

### CODE REVIEW

File: src/main/java/org/dcsa/bkg/service/impl/BookingServiceImpl.java [hunk 1]
```diff
@@ -181,6 +181,10 @@ private Pageable mapSortParameters(Pageable pageable) {
               "The attribute exportDeclarationReference cannot be null if isExportDeclarationRequired is true."));
     }
 
+    if (bookingRequest.getVesselIMONumber() == null && bookingRequest.getExportVoyageNumber() == null && bookingRequest.getExpectedDepartureDate() == null) {
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/bkg/service/impl/BookingServiceImpl.java:184]: You are forgetting the arrival date in this check - if the above 3 fields need to be combined with arrivalDate = null

File: src/main/java/org/dcsa/bkg/service/impl/BookingServiceImpl.java [hunk 2]
```diff
@@ -181,6 +181,14 @@ private Pageable mapSortParameters(Pageable pageable) {
...
     }
 
+    if (bookingRequest.getExpectedArrivalDateStart() == null && bookingRequest.getExpectedArrivalDateStart() == null && bookingRequest.getExpectedDepartureDate() == null) {
+        return Mono.error(new CreateException("The attribute expectedDepartureDate cannot be null if both expectedArrivalDateStart and expectedArrivalDateEnd are null."));
+    }
+
+    if (bookingRequest.getVesselIMONumber() == null && bookingRequest.getExportVoyageNumber() == null && bookingRequest.getExpectedDepartureDate() == null) {
+        return Mono.error(new CreateException("The attribute expectedDepartureDate cannot be null if vesselIMONumber/exportVoyageNumber is null."));
+    }
+
```
Comment (HenrikHL) [inline src/main/java/org/dcsa/bkg/service/impl/BookingServiceImpl.java:191]:
```suggestion
    if (bookingRequest.getExpectedArrivalDateStart() == null && bookingRequest.getExpectedArrivalDateStart() == null && bookingRequest.getExpectedDepartureDate() == null && bookingRequest.getVesselIMONumber() == null && bookingRequest.getExportVoyageNumber() == null) {
        return Mono.error(new CreateException("The attributes expectedArrivalDateStart, expectedArrivalDateEnd, expectedDepartureDate and vesselIMONumber/exportVoyageNumber cannot all be null at the same time. These fields are conditional and require that at least one of them is not empty."));
    }

```

Comment (HenrikHL) [reply src/main/java/org/dcsa/bkg/service/impl/BookingServiceImpl.java:191]: May I suggest to consolidate this into 1 condition

File: src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java [hunk 3]
```diff
@@ -450,7 +456,43 @@ void testCreateBookingWhenIsImportLicenseRequiredIsTrueAndImportLicenseReference
...
+                        Assertions.assertTrue(throwable instanceof CreateException);
+                        assertEquals(
+                                "The attribute expectedDepartureDate cannot be null if both expectedArrivalDateStart and expectedArrivalDateEnd are null.",
+                                throwable.getMessage());
+                      })
+              .verify();
+    }
+
+    @Test
+    @DisplayName("Method should throw an exception when vesselIMONumber, exportVoyageNumber, and expectedDepartureDate are null")
```
Comment (HenrikHL) [inline src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java:477]: Same comment as above - just vice-versa

File: src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java [hunk 4]
```diff
@@ -450,7 +456,43 @@ void testCreateBookingWhenIsImportLicenseRequiredIsTrueAndImportLicenseReference
     }
 
     @Test
-    @DisplayName("Method should save and return shallow booking for given booking request")
+    @DisplayName("Method should throw an exception when expectedArrivalDateStart, expectedArrivalDateEnd, and expectedDepartureDate are null")
```
[the remark this answers is not in the document]
Comment (mp-assco) [reply src/test/java/org/dcsa/bkg/service/impl/BookingServiceImplTest.java]: Test cases have been updated

File: src/main/java/org/dcsa/bkg/model/transferobjects/AbstractBookingTO.java [hunk 5]
```diff
@@ -5,6 +5,7 @@
 import lombok.Data;
 import org.dcsa.core.events.model.enums.*;
 import org.dcsa.core.validator.ValidVesselIMONumber;
+import org.springframework.data.relational.core.mapping.Column;
```
Comment (douradofrazer) [inline src/main/java/org/dcsa/bkg/model/transferobjects/AbstractBookingTO.java:8]: Not needed.

**Review by douradofrazer — DISMISSED** [review]
LGTM

**Review by douradofrazer — APPROVED** [review]

**Pull request merged**